File: d20/main.py
# Copy from: https://github.com/Schellcunn/Advent2024/blob/main/Day20/race2.py
import time
from collections import deque

START = "S"
END = "E"
WALL = "#"


def inWorld(row:int,col:int,track:list[list[str]]):
    return 0 <= row < len(track) and 0 <= col < len(track[row])

def bfs(track:list[list[str]], start:tuple[int,int], end:tuple[int,int]):
    """
    BFS to return one path from start to end (list of (row,col) positions).
    If no path found, returns an empty list.
    """
    from collections import deque
    queue = deque()
    queue.append(start)
    visited = {start}
    parent = {start: None}  # Track predecessors for path reconstruction

    while queue:
        row, col = queue.popleft()
        if (row, col) == end:
            # Reconstruct path from end to start
            path = []
            cur = end
            while cur is not None:
                path.append(cur)
                cur = parent[cur]
            return path[::-1]  # Reverse it to go from start -> end
        for dr, dc in [(0,1),(0,-1),(1,0),(-1,0)]:
            nr, nc = row + dr, col + dc
            if not inWorld(nr, nc, track):
                continue
            if track[nr][nc] == WALL:
                continue
            if (nr, nc) not in visited:
                visited.add((nr, nc))
                parent[(nr, nc)] = (row, col)
                queue.append((nr, nc))
    return []

def getCheats(path:list[tuple[int,int]], maxCheats:int, threshold:int):
    """
    path: BFS-discovered path from S -> E as [(r,c), (r,c), ...].
    maxCheats: max distance to 'skip' with cheat
    threshold: required amount of time saved
    """
    raceTime = len(path) - 1  # total steps
    count = 0
    for i in range(len(path)-1):
        for j in range(i+1, len(path)):
            posA, posB = path[i], path[j]
            distance = abs(posA[0] - posB[0]) + abs(posA[1] - posB[1])
            if distance > maxCheats:
                continue
            cheatTime = (i + 1) + distance + (raceTime - j - 1)
            timeSaved = raceTime - cheatTime
            if timeSaved >= threshold:
                count += 1
    print(count)

def getTrack(fileName:str):
    with open(fileName, 'r') as file:
        track = []
        start = (0,0)
        end = (0,0)
        for i, line in enumerate(file):
            rowData = line.strip()
            if not rowData:
                continue
            curRow = []
            for j, ch in enumerate(rowData):
                curRow.append(ch)
                if ch == END:
                    end = (i,j)
                elif ch == START:
                    start = (i,j)
            track.append(curRow)
        return track, start, end

def getBestCheatCounts(fileName:str, maxCheats:int, threshold:int):
    track, start, end = getTrack(fileName)
    path = bfs(track, start, end)
    if not path:
        print(f"No path found in {fileName}.")
    else:
        getCheats(path, maxCheats, threshold)

def main():
    # Single test with input3.txt
    print("Processing input3.txt...")
    getBestCheatCounts("input.txt", 20, 100)

if __name__ == "__main__":
    main()

File: d20/main.go
package main

import (
	"fmt"
	"github.com/swh00tw/aoc"
	"math"
)

var fileName = "input.txt"

func loadInput() [][]byte {
	lines := aoc.LoadInputLines(fileName)
	matrix := make([][]byte, len(lines))
	for i, line := range lines {
		matrix[i] = []byte(line)
	}
	return matrix
}

func BFS(x, y int, matrix [][]byte) [][]int {
	m := len(matrix)
	n := len(matrix[0])
	visited := make([][]int, m)
	for i := range visited {
		visited[i] = make([]int, n)
		for j, _ := range visited[i] {
			visited[i][j] = math.MaxInt
		}
	}

	queue := make([][]int, 0)
	queue = append(queue, []int{x, y})
	visited[x][y] = 0

	for len(queue) > 0 {
		// pop
		curr := queue[0]
		queue = queue[1:]
		x, y := curr[0], curr[1]

		for _, dir := range [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}} {
			nx, ny := x+dir[0], y+dir[1]
			if nx >= 0 && nx < m && ny >= 0 && ny < n && matrix[nx][ny] != '#' && visited[nx][ny] == math.MaxInt {
				visited[nx][ny] = visited[x][y] + 1
				queue = append(queue, []int{nx, ny})
			}
		}
	}

	return visited
}

func getStartEnd(matrix [][]byte) [][]int {
	start := make([]int, 2)
	end := make([]int, 2)
	for i, row := range matrix {
		for j, cell := range row {
			if cell == 'S' {
				start[0], start[1] = i, j
			} else if cell == 'E' {
				end[0], end[1] = i, j
			}
		}
	}
	return [][]int{start, end}
}

func solve(matrix [][]byte, threshold int, cheatAllowed int) int {
	startEnd := getStartEnd(matrix)
	start, _ := startEnd[0], startEnd[1]
	visited := BFS(start[0], start[1], matrix)
	m := len(matrix)
	n := len(matrix[0])

	save := make(map[int]aoc.Set[string])

	pathKey := func(x1, y1, x2, y2 int) string {
		return fmt.Sprintf("%d,%d,%d,%d", x1, y1, x2, y2)
	}

	// for each cell on the path
	for i, row := range matrix {
		for j, cell := range row {
			if cell == '.' || cell == 'S' {
				visitedCheat := aoc.Copy2DArray(visited)

				queue := make([][]int, 0)
				queue = append(queue, []int{i, j, 0}) // x, y, iter
				for len(queue) > 0 {
					// pop
					curr := queue[0]
					queue = queue[1:]
					x, y, iter := curr[0], curr[1], curr[2]
					if iter > 0 && (matrix[x][y] == 'E' || matrix[x][y] == '.') {
						diff := visited[x][y] - visitedCheat[x][y]
						if _, ok := save[diff]; !ok {
							save[diff] = aoc.Set[string]{}
						}
						key := pathKey(x, y, i, j)
						save[diff].Add(key)
					}
					if matrix[x][y] == 'E' {
						continue
					}

					for _, dir := range [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}} {
						nx, ny := x+dir[0], y+dir[1]
						if nx >= 0 && nx < m && ny >= 0 && ny < n {
							if visitedCheat[x][y]+1 <= visitedCheat[nx][ny] && iter < cheatAllowed {
								visitedCheat[nx][ny] = visitedCheat[x][y] + 1
								queue = append(queue, []int{nx, ny, iter + 1})
							}
						}
					}
				}

			}
		}
	}

	ans := make(map[int]int)
	for k, v := range save {
		if k >= threshold {
			ans[k] = v.Len()
		}
	}
	fmt.Println("ans: ", ans)

	cnt := 0
	for _, v := range ans {
		cnt += v
	}

	return cnt
}

func main() {
	matrix := loadInput()

	fmt.Println("Part 1: ", solve(matrix, 1, 2))
	fmt.Println("Part 2: ", solve(matrix, 100, 20))
}
